### PR TITLE
add --with-system-boost configure option for using system-wide boost headers

### PR DESCRIPTION
**Pull request by real-dam**
This is the last one of the "use system $lib" series.

### CODE REVIEW

File: src/include/firebird/Message.h [hunk 1]
```diff
@@ -25,7 +25,15 @@
 
 #include "ibase.h"
 #include "./Interface.h"
+#if defined USE_SYSTEM_BOOST
```
Comment (asfernandes) [inline src/include/firebird/Message.h:28]: This is public header file. So I think name must be better `FB_USE_SYSTEM_BOOST`.

**Comment by aafemt** [comment]
Taking into account that C++11 features are fully available, do we still need boost?

**Comment by real-dam** [comment]
> Taking into account that C++11 features are fully available, do we still need boost?

I like the idea of replacing an external dependency with a compiler feature. Hopefully someone more fluent in C++ than me would try it.

**Comment by asfernandes** [comment]
> Taking into account that C++11 features are fully available, do we still need boost?

We didn't discussed C++11 in public headers yet. I support it.

File: configure.ac [hunk 2]
```diff
@@ -560,6 +560,12 @@ AC_ARG_WITH(system-re2,
   [RE2_BUILD=N])
 AC_SUBST(RE2_BUILD)
 
+LOCAL_BOOST=Y
+AC_ARG_WITH(system-boost,
+  [  --with-system-bools     use system-wide boost library instead of embedded copy],
```
Comment (asfernandes) [inline configure.ac:565]: This is misspelled (bools).

**Comment by asfernandes** [comment]
How that will work for public headers?

Will you put a dependency for boost in firebird-client-dev (or the name you use) package?

Will you deploy fb-boost headers in that package?

Will the user also need to use the define?

**Comment by real-dam** [comment]
> How that will work for public headers?
> 
> Will you put a dependency for boost in firebird-client-dev (or the name you use) package?
> 
> Will you deploy fb-boost headers in that package?
> 
> Will the user also need to use the define?

Good question. Requiring that all consumers of `firebird/Message.h` define `USE_SYSTEM_BOOST` would be bad.

Perhaps I should reverse the define and rename it to `USE_FB_BOOST`? This way consumers using the Debian package would only need to have boost around, maybe pulled by a dependency.

That dependency, while technically needed, is not present in firebird3.0 packages which have a similar patch (crudely using system boost and replacing `FB_BOOST_xxx` with `BOOST_xxx` in `Message.h`) and nobody complained for the last 5+ years. I guess some softer form of dependency would be enough.

Will update the PR with `USE_FB_BOOST` and reverse meaning.

**Comment by aafemt** [comment]
> nobody complained for the last 5+ years.

It just means that nobody uses Messages.h.

**Comment by real-dam** [comment]
> Will update the PR with `USE_FB_BOOST` and reverse meaning.
Oh, but that would mean that any `Message.h` consumers that have it from non-packaged Firebird would need the define. Not good either.

I think the approach used in fb3 Debian packages is the least evil.

The cleanest way from consumer side would be to have `Messages.h` with either `include <...boost...>` (plus a dependency) or `include "...boost..."`, both hard-coded and not controlled by a define.

fb3 debs took one approach at this, patching the installed files.

An approach that controls the output via `--with-system-boost` would require similar sort of patching, for example via sed. Shouldn't be a problem since this affects only posix builds. Will be kind of ugly, but should work.

**Comment by real-dam** [comment]
This became as good as patching during package build with the added benefit that including it upstream will make the functionality available for other downstream distributions.

**Pull request merged**